src/cryptonote_core/service_node_swarm.h

```diff
@@ -0,0 +1,33 @@
+#pragma once
+
+#include "service_node_rules.h"
+
+#include <map>
+#include <vector>
+#include <random>
+
+namespace service_nodes {
+    using swarm_snode_map_t = std::map<swarm_id_t, std::vector<crypto::public_key>>;
+    struct swarm_size {
+        swarm_id_t swarm_id;
+        size_t size;
+    };
+    struct excess_pool_snode {
+        crypto::public_key public_key;
+        swarm_id_t swarm_id;
+    };
+
+    void calc_swarm_changes(swarm_snode_map_t& swarm_to_snodes, uint64_t seed);
+
+#ifdef UNIT_TEST
+    size_t calc_excess(const swarm_snode_map_t &swarm_to_snodes);
+    size_t calc_threshold(const swarm_snode_map_t &swarm_to_snodes);
+    crypto::public_key steal_from_excess_pool(swarm_snode_map_t &swarm_to_snodes, std::mt19937_64 &mt);
+    void create_new_swarm_from_excess(swarm_snode_map_t &swarm_to_snodes, std::mt19937_64 &mt);
+    void calc_swarm_sizes(const swarm_snode_map_t &swarm_to_snodes, std::vector<swarm_size> &sorted_swarm_sizes);
+    void assign_snodes(const std::vector<crypto::public_key> &snode_pubkeys, swarm_snode_map_t &swarm_to_snodes, std::mt19937_64 &mt);
+    void get_excess_pool(size_t threshold, swarm_snode_map_t& swarm_to_snodes, std::vector<excess_pool_snode>& pool_snodes, size_t& excess);
+    const excess_pool_snode& pick_from_excess_pool(const std::vector<excess_pool_snode>& excess_pool, std::mt19937_64 &mt);
+    bool remove_excess_snode_from_swarm(const excess_pool_snode& excess_snode, swarm_snode_map_t &swarm_to_snodes);
```

Comment:
What does this return (need more comments)? The return is mostly ignored and only used in unit tests. Should we move the logic of checking the new size to the tests?

Reply:
mmh yeah I wasn't super happy with this either.
the `bool` returned indicates if the snode was removed from the swarm map. Maybe I should use the return value wherever I use that function?

Reply:
I wonder if it is necessary. This function could just have a precondition that the element to remove is in the map (use assert?). In your code it is always going to be the case.

Reply:
Ok I will change it to return `void` and get the unit test to check for deletion.

Reply:
Actually, you are already using `map::at` which would throw if the element is not found, so you would never return false in this case.